add_keys.py

```diff
@@ -0,0 +1,204 @@
+        pool_address = Web3.toChecksumAddress(pool_address)
+
+    pool_abi_path = os.path.join(ARTIFACTS_DIR, POOL_ARTIFACT_FILE)
+    registry_abi_path = os.path.join(ARTIFACTS_DIR, REGISTRY_ARTIFACT_FILE)
+
+    # Get Pool contract
+    with open(pool_abi_path, 'r') as file:
+        abi = json.load(file)
+        file.close()
+    pool = w3.eth.contract(abi=abi['abi'], address=pool_address)  # contract object
+
+    # Get Registry contract
+    registry_address = pool.functions.getOperators().call()
+    logging.info(f'{registry_address=}')
+
+    with open(registry_abi_path, 'r') as file:
+        abi = json.load(file)
+        file.close()
+    return w3.eth.contract(abi=abi['abi'], address=registry_address)
+
+
+def init_account(w3: Web3):
+    '''
+    Initialise the signer.
+    '''
+    privkey = os.getenv('PRIV_KEY')
+    account = w3.eth.account.from_key(privkey)
+    logging.info(f'Account: {account.address}')
+    return account
+
+
+
+@click.command()
+@click.option("--nonce", default=-1, type=int, help="Nonce to use (default uses next nonce on chain)")
+@click.option("--gas", default=GAS_LIMIT, type=int, help="Gas limit - defaults to {}, which is suitable for submitting 100 keys.".format(GAS_LIMIT))
+@click.option("--operator", help="Operator ID.", type=int)
+@click.option("--eth1-uri", help="Ethereum node address", type=str)
+@click.option("--pool-address", help="Lido pool contract address - defaults to mainnet contract address", type=str, default=POOL_ADDRESS)
+@click.argument('filename')
+def main(filename, nonce, operator, gas, eth1_uri, pool_address):
+    '''
+    Main entrypoint.
+    '''
+    try:
+        if eth1_uri == None:
+            eth1_uri = os.environ.get('WEB3_PROVIDER_URI', None)
+            if eth1_uri == None:
+                raise Exception("Must pass --eth1-uri flag or set WEB3_PROVIDER_URI environment variable")
+        w3 = init_provider(eth1_uri)
+        registry = init_contracts(w3, pool_address)
+        account = init_account(w3)
+
+        if operator is None:
+            raise Exception("--operator flag is required")
+
+        with open(filename, 'r') as f:
+            keyfile = json.load(f)
+            f.close()
+
+        keys = [x.get('pubkey') for x in keyfile]
+        signatures = [x.get('signature') for x in keyfile]
+        logging.info("Found %d keys...", len(keys))
+
+        tx = build_tx(registry, operator, keys, signatures, account, gas)
+        send_tx(w3, tx, account, nonce)
+    except Exception as e:
+        print("Error: {}".format(e))
+        sys.exit(1)
+
+
+def build_tx(registry, operator_id, keys, signatures, account, gas):
+    '''
+    Build an addSigningKeysOperatorBH tx from the keys and signatures provided.
+    '''
+
+    # sanity_check
+    if len(keys) != len(signatures):
+        raise Exception("Imbalance in keys and signatures")
+
+    return registry.functions.addSigningKeysOperatorBH(
+        operator_id,
+        len(keys),
+        "".join(keys),
```

Comment:
I don’t know how this `addSigningKEysOperatorBH` works and what the type of `key` is, but I suppose `keys: List[str]`. If we concatenate them without separator, then how does `addSigningKEysOperatorBH` tell them apart? Are they fixed-length? If keys contains base58, base58 is not a fixed-length encoding (i.e. the size of the output depends on the input *data*, not just on the input length), so that would be tricky.

Reply:
So this method come from the web3 library reading the abi.json file that the smart contract was created with. The addresses are 32 byte hex strings, so are fixed length. So keys is a list of hex strings (although addSigningKeysOperatorBH takes []byte (or whatever the python version of this is).

Reply:
Ah, I see, then concatenating them sounds safe :+1: